src/engine/LocalVocab.h

```diff
@@ -0,0 +1,69 @@
+// Copyright 2022, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Author: Hannah Bast <bast@cs.uni-freiburg.de>
+
+#pragma once
+
+#include "absl/container/node_hash_map.h"
+
+// A class for maintaing a local vocabulary with contiguous (local) IDs. This is
+// meant for words that are not part of the normal vocabulary (constructed from
+// the input data at indexing time).
+//
+// TODO: This is a first version of this class with basic functionality. Note
+// that the local vocabulary used to be a simple `std::vector<std::string>`
+// defined inside of the `ResultTable` class. You gotta start somewhere.
+class LocalVocab {
+ public:
+  // Create a new, empty local vocabulary.
+  LocalVocab() = default;
+
+  // Prevent accidental copying of a local vocabulary (it can be quite large),
+  // but moving it is OK.
+  //
+  // TODO: does the default move do the "right" thing, that is, move the hash
+  // map instead of copying it?
+  LocalVocab(const LocalVocab&) = delete;
+  LocalVocab(LocalVocab&&) = default;
+
+  // Get ID of a word in the local vocabulary. If the word was already
+  // contained, return the already existing ID. If the word was not yet
+  // contained, add it, and return the new ID.
+  Id getIdAndAddIfNotContained(const std::string& word);
+
+  // The number of words in the vocabulary.
+  size_t size() const { return idsToWordsMap_.size(); }
+
+  // Return true if and only if the local vocabulary is empty.
+  bool empty() const { return idsToWordsMap_.empty(); }
+
+  // Return a const reference to the word.
+  const std::string& getWord(LocalVocabIndex localVocabIndex) const;
```

Comment:
If nothing breaks in the rest of the code, I would return a `string_view` here,
this makes us more agnostic against the implementation in the future
(no strong opinion though, depends on the further implementation).

Reply:
Can you give an example use of this function, where it would be an advantage if it returned a string_view?

Reply:
The benefit is not that great if we are relatively sure that we store `std::string*` for space reasons and will continue to do so, so we can leave it as is.